streams/test-utils/src/main/java/org/apache/kafka/streams/test/MockTime.java

```diff
@@ -0,0 +1,130 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.test;
+
+import org.apache.kafka.common.utils.Time;
+
+import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A clock that you can manually advance by calling {@code sleep()}.
+ */
+public class MockTime implements Time {
+
+    /**
+     * Listener that is invoked each time <em>after</em> time was advanced or modified.
+     */
+    public interface MockTimeListener {
+        void tick(final long currentTimeMs);
+    }
+
+    private final CopyOnWriteArrayList<MockTimeListener> listeners = new CopyOnWriteArrayList<>();
+
+    private long autoTickMs;
+
+    // Values from `nanoTime` and `currentTimeMillis` are not comparable, so we store them separately to allow tests
+    // using this class to detect bugs where this is incorrectly assumed to be true
+    private final AtomicLong timeMs;
+    private final AtomicLong highResTimeNs;
+
+    public MockTime() {
+        this(0);
+    }
+
+    public MockTime(final long autoTickMs) {
+        this(autoTickMs, System.currentTimeMillis(), System.nanoTime());
+    }
+
+    public MockTime(final long autoTickMs,
+                    final long currentTimeMs,
+                    final long currentHighResTimeNs) {
+        if (autoTickMs < 0) {
+            throw new IllegalArgumentException("autoTickMs cannot be negative.");
+        }
+        this.timeMs = new AtomicLong(currentTimeMs);
+        this.highResTimeNs = new AtomicLong(currentHighResTimeNs);
+        this.autoTickMs = autoTickMs;
+    }
+
+    public void addListener(final MockTimeListener listener) {
+        Objects.requireNonNull(listener);
+        listeners.add(listener);
+    }
+
+    @Override
+    public long milliseconds() {
+        final long currentTime = timeMs.get();
+        maybeSleep(autoTickMs);
+        return currentTime;
+    }
+
+    @Override
+    public long nanoseconds() {
+        final long currentTime = highResTimeNs.get();
+        maybeSleep(autoTickMs);
+        return currentTime;
+    }
+
+    @Override
+    public long hiResClockMs() {
+        return TimeUnit.NANOSECONDS.toMillis(nanoseconds());
+    }
+
+    private void maybeSleep(final long ms) {
+        if (ms != 0)
+            sleep(ms);
+    }
+
+    @Override
+    public void sleep(final long ms) {
```

Comment:
IMHO `sleep` doesn't feel quite right, what about SMT along the lines of `advance` or `process`

Reply:
In the mock impl nothing is paused, but to the runtime it's just `Time` object that could build on `SystemTime` and thus really sleep here. We could add a second method `advance()` that internally just calls `sleep()` though to make tests more readable. But having two methods doing the same thing might be confusion, too. Thus, I prefer to leave as is. WDYT?

Reply:
Yeah, I'd have to agree it doesn't make sense to have two methods doing the same thing, I'd say leave as is.